sootup.java.bytecode/src/main/java/sootup/java/bytecode/frontend/AsmUtil.java

```diff
@@ -154,60 +154,96 @@ public static Type toJimpleSignature(@Nonnull String desc) {
 
   @Nonnull
   public static Type toJimpleType(@Nonnull String desc) {
-    int idx = desc.lastIndexOf('[');
-    int nrDims = idx + 1;
+    int nrDims = countArrayDim(desc);
     if (nrDims > 0) {
-      if (desc.charAt(0) != '[') {
+      desc = desc.substring(nrDims);
+    }
+
+    Type baseType = toPrimitiveOrVoidType(desc).orElse(null);
+    if (baseType == null && desc.charAt(0) != 'L') {
+      throw new AssertionError("Unknown descriptor: " + desc);
+    }
+
+    if (baseType == null && desc.charAt(0) == 'L') {
+      if (desc.charAt(desc.length() - 1) != ';') {
+        throw new AssertionError("Invalid reference descriptor: " + desc);
+      }
+      String name = desc.substring(1, desc.length() - 1);
+      name = toQualifiedName(name);
+      baseType = JavaIdentifierFactory.getInstance().getType(toQualifiedName(name));
+    }
+    if (!(baseType instanceof JavaClassType) && desc.length() > 1) {
+      throw new AssertionError("Invalid primitive type descriptor: " + desc);
+    }
+    return nrDims > 0
+        ? JavaIdentifierFactory.getInstance().getArrayType(baseType, nrDims)
+        : baseType;
+  }
+
+  @Nonnull
+  public static Type ArrayTypetoJimpleType(@Nonnull String desc) {
+    int nrDims = countArrayDim(desc);
+    if (nrDims > 0) {
+      desc = desc.substring(nrDims);
+    }
+
+    Type baseType = toPrimitiveOrVoidType(desc).orElse(null);
+
+    if (baseType == null) {
+      // Objects can be either Ljava/lang/Object; or java/lang/Object
+      String name = desc;
+      if (desc.charAt(0) == 'L' && desc.charAt(desc.length() - 1) != ';') {
+        name = desc.substring(1, desc.length() - 1);
+      }
+      baseType = JavaIdentifierFactory.getInstance().getType(toQualifiedName(name));
+    }
+
+    return nrDims > 0
+        ? JavaIdentifierFactory.getInstance().getArrayType(baseType, nrDims)
+        : baseType;
+  }
+
+  /** returns the amount of dimensions of a description. */
+  private static int countArrayDim(@Nonnull String desc) {
+    int nrDims = desc.lastIndexOf('[') + 1;
+    for (int index = 0; index < nrDims; index++) {
+      if (desc.charAt(index) != '[') {
         throw new AssertionError("Invalid array descriptor: " + desc);
       }
-      desc = desc.substring(idx + 1);
     }
-    Type baseType;
+    return nrDims;
```

Comment:
is the format check necessary? (i know it was there before)

Reply:
Before the change, the check was only the last [ and the first [
I think the question is if we want to be robust.
This method is normally called with descriptions extracted from ASM, so there shouldnt be an issue with the format.
But this is also a public method so you could feed it with selfcreated descprictions.